### PR TITLE
[FLINK-16480] Improve error reporting on AZP

### PR DESCRIPTION
**Pull request by rmetzger**
## What is the purpose of the change

While debugging unstable e2e tests and crashing JVMs remotely, I came up with a bunch of improvements to our testing scripts, which I would like to add to the main repo.

## Brief change log

The following changes are contained here:
- [FLINK-16480][e2e] Always checks logs when test has failed: The e2e tests now check the logs for exceptions / errors etc when a test has failed (before, logs where not checked in error cases for some tests)
- [FLINK-16480][AZP] fix log upload condition the log upload failed for e2e tests, when a bash e2e test failed.
- [FLINK-16480] Collect debug files when JVM crashes (for IT cases): we now collect coredumps, jvm stacktraces, and other debugging files
-  [FLINK-16480][e2e] Collect java crashreports for e2e tests as well: same for e2e tests. This includes a refactoring of the log collection tools into a separate file.

## Verifying this change

This run properly collects coredumps etc of a test that used to fail on master a few days ago: https://dev.azure.com/rmetzger/Flink/_build/results?buildId=6137&view=results

## Does this pull request potentially affect one of the following parts:

affects only tests

### CODE REVIEW

**Comment by rmetzger** [comment]
There might be some interference with the common_utils.sh / on_exit trap handling that I am currently looking into

**Comment by rmetzger** [comment]
I validated and there is no interference.

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh [hunk 1]
```diff
@@ -46,7 +46,7 @@ function run_test {
 
     function test_error() {
       echo "[FAIL] Test script contains errors."
-      post_test_validation 1 "$description" "$skip_check_exceptions"
+      post_test_validation 1 "$description"
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test-runner-common.sh]: do you know whether the bash tests upload the logs for failed tests?

File: tools/travis/common-logging.sh [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	ARTIFACTS_FILE=${TRAVIS_JOB_NUMBER}.tar.gz
+	if [ ! -z "$TF_BUILD" ] ; then
+		# set proper artifacts file name on Azure Pipelines
+		ARTIFACTS_FILE=${BUILD_BUILDNUMBER}.tar.gz
+		if [ ! -z "$MODULE" ] ; then
+			ARTIFACTS_FILE=${BUILD_BUILDNUMBER}-$(echo $MODULE | tr -dc '[:alnum:]\n\r').tar.gz
+		fi
+	fi
+
+	echo "PRODUCED build artifacts."
```
Comment (zentol) [inline tools/travis/common-logging.sh]: as usual this would be easier to review if the moving of code were put into a separate commit.

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh (hunk 1 above)
Comment (zentol) [reply flink-end-to-end-tests/test-scripts/test-runner-common.sh]: My impressions is that this will actually obfuscate errors. People pretty much only look at the last ~20 lines of the CI output, and don't properly investigate the case, resulting in several superfluous JIRAs.

File: tools/travis_watchdog.sh [hunk 3]
```diff
@@ -177,6 +177,14 @@ print_stacktraces () {
 	done
 }
 
+collect_coredumps() {
+	echo "Searching for .dump, .dumpstream and related files in $($HERE/../)"
+	for file in `find . -type f -regextype posix-extended -iregex '.*\.dump|.*\.dumpstream|.*hs.*\.log|.*/core(.[0-9]+)?$'`; do
```
Comment (zentol) [inline tools/travis_watchdog.sh]: a simpler solution may be to setup the JVM to directly write the dumps into `ARTIFACTS_DIR`.
https://stackoverflow.com/questions/1880166/is-it-possible-to-specify-where-jvms-crash-dumps-go

File: tools/azure_controller.sh [hunk 4]
```diff
@@ -53,6 +51,8 @@ print_system_info() {
 
 print_system_info
 
+# enable core dumps
+sudo ulimit -c unlimited
```
Comment (zentol) [inline tools/azure_controller.sh]: what does this have to do with core dumps?

File: tools/azure_controller.sh [hunk 5]
```diff
@@ -24,8 +24,6 @@ echo $MAVEN_OPTS
 mvn -version
 echo "Commit: $(git rev-parse HEAD)"
 
-
-
```
Comment (zentol) [inline tools/azure_controller.sh]: unrelated

File: tools/travis/common-logging.sh [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+# These variables are stored as secure variables in '.travis.yml', which are generated per repo via
+# the travis command line tool.
+UPLOAD_BUCKET=$ARTIFACTS_AWS_BUCKET
+UPLOAD_ACCESS_KEY=$ARTIFACTS_AWS_ACCESS_KEY
+UPLOAD_SECRET_KEY=$ARTIFACTS_AWS_SECRET_KEY
+
+
+SCRIPT_DIR="`dirname \"$0\"`"
+SCRIPT_DIR="`( cd \"${SCRIPT_DIR}\" && pwd -P)`"
+export FLINK_ROOT="`( cd \"${SCRIPT_DIR}/..\" && pwd -P)`"
```
Comment (zentol) [inline tools/travis/common-logging.sh]: this seems quite brittle

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh (hunk 1 above)
Comment (rmetzger) [reply flink-end-to-end-tests/test-scripts/test-runner-common.sh]: > do you know whether the bash tests upload the logs for failed tests?

No. All directories are cleaned up after a test (in any case).

> My impressions is that this will actually obfuscate errors. People pretty much only look at the last ~20 lines of the CI output, and don't properly investigate the case, resulting in several superfluous JIRAs.

That is definitely a risk, I agree. But I would prefer not to rework this change for a problem we have not confirmed to exist.
It is quite tedious to improve these things because of all the testing cycles in between.
I would prefer to add this as is and re-adjust later if we have the feeling that people are not checking the logs carefully enough.

File: tools/azure_controller.sh (hunk 4 above)
Comment (rmetzger) [reply tools/azure_controller.sh]: https://stackoverflow.com/questions/28982396/failed-to-write-core-dump-core-dumps-have-been-disabled

I guess the default core dump size is too small or even 0.

File: tools/travis_watchdog.sh (hunk 3 above)
Comment (rmetzger) [reply tools/travis_watchdog.sh]: With this implementation, we can use this function also for collecting any JVM debugging files from the end to end tests.
Yes, we could also inject JVM properties through environment variables into most of the end to end tests. But who guarantees that Kafka, ElasticSearch, Zookeeper all properly respect the environment variables.

With this approach, I'm following a blackbox approach where I don't trust the system I'm testing.

Also, I don't know if I would spend my time well for the project if I would refactor something that works. Coredumps (luckily) happen quite rarely. I just wanted to make life for others easier when they have to debug a crashing JVM on a CI system.

File: tools/azure_controller.sh (hunk 5 above)
Comment (rmetzger) [reply tools/azure_controller.sh]: You are right. I need to read the guidelines more closely.

File: tools/travis/common-logging.sh (hunk 2 above)
Comment (rmetzger) [reply tools/travis/common-logging.sh]: That's what I tried to do. There's a commit introducing core dump collection for regular tests, and another commit for introducing it for end to end tests.

I could not find any guidelines on how to exactly split the commits. How do you want it?

File: tools/travis/common-logging.sh (hunk 6 above)
Comment (rmetzger) [reply tools/travis/common-logging.sh]: I agree. Do you have a good idea how to make it more stable?

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh (hunk 1 above)
Comment (zentol) [reply flink-end-to-end-tests/test-scripts/test-runner-common.sh]: > But I would prefer not to rework this change for a problem we have not confirmed to exist.

We have repeatedly seen people not properly analyzing the CI output when the logs are being dumped, this is not a theory.

> It is quite tedious to improve these things because of all the testing cycles in between.

That comes with the territory unfortunately. You will just have to find a was to reduce these cycles, e.g., only compile and test flink-core or refactor the scripts such that you can run things offline.

File: tools/travis_watchdog.sh (hunk 3 above)
Comment (zentol) [reply tools/travis_watchdog.sh]: > But who guarantees that Kafka, ElasticSearch, Zookeeper all properly respect the environment variables.

There's no guarantee that these are run under FLINK_ROOT though; the java e2e tests use a junit temporary folder to store and run these services, as well as the flink distribution.
At least you'd also have to search the entire tmp directory of the system, or modify `java.io.tmpdir` to point somewhere under FLINK_ROOT.

Have you tried the `JAVA_TOOL_OPTIONS` environment variable that the JVM itself evaluates? https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/envvars002.html

> I don't know if I would spend my time well for the project if I would refactor something that works.

This can be read in rather frightening ways: "Lol let's use the first version that works _now_."

File: tools/travis/common-logging.sh (hunk 6 above)
Comment (zentol) [reply tools/travis/common-logging.sh]: AFAIK you can pass parameters when calling `source`, so you could explicitly pass in the desired directory.

File: tools/azure_controller.sh [hunk 7]
```diff
@@ -54,7 +54,7 @@ print_system_info() {
 print_system_info
 
 # enable core dumps
-sudo ulimit -c unlimited
+ulimit -c unlimited
```
Comment (zentol) [inline tools/azure_controller.sh]: I suppose the sudo removal belongs into another commit.

File: tools/travis/common-logging.sh (hunk 2 above)
Comment (zentol) [reply tools/travis/common-logging.sh]: From what I can tell you first moved a bunch of stuff from `travis_watchdog.sh` to `common-logging.sh` (commit#1, pure refactoring, no behavioral changes), and then modified `run-nightly-tests.sh` to make use of it (commit#2, behavioral changes).

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh (hunk 1 above)
Comment (rmetzger) [reply flink-end-to-end-tests/test-scripts/test-runner-common.sh]: Okay. I will look into a different approach for making the logs available.

**Comment by rmetzger** [comment]
I'm closing this PR, re-opening parts of this PR as separate PRs.
